File: Homework 2/problem1.py
input_size = 10

#A simple print function
def print_arr(input):
    for i in range(input_size):
        print (input[i])

#The partition function
def partition(input,p,r):
    pivot = input[r];
    while p < r:
        while input[p] < pivot:
            p = p + 1

        while input[r] > pivot:
            r = r -1
        if input[p] == input[r]:
            p = p + 1
        elif p < r:
            tmp = input[p]
            input[p] = input[r]
            input[r] = tmp
    return r

#The quicksort recursive function
def quickSort(input, p, r):
    if p < r:
        j = partition(input, p, r)
        quickSort(input, p, j-1) 
        quickSort(input, j+1, r) 


input = [500, 700, 800, 100, 300, 200, 900, 400, 1000, 600]

print ("Input: ")
print_arr(input)

quickSort(input, 0, input_size-1)

print ()
print ("Output: ") 
print_arr(input)

File: Homework 4/problem1.java
package problem1;

public class quicksort {
	final static int INPUT_SIZE = 10;
	
	public static void main(String[] args) {
		int[] input = {500, 700, 800, 100, 300, 200, 900, 400, 1000, 600};
		System.out.println("Input: ");
		print(input);
		quicksort1(input, 0 ,9);
		System.out.println("Output: ");
		print(input);
	}
	
	public static void print(int[] input) {
		for (int i=0; i<INPUT_SIZE; i++) {
			System.out.println(input[i] + " ");
		}
		System.out.println("\n");
	}
	
	public static void quicksort1(int[] input, int p, int r) {
		if (p < r) {
			int j = partition(input, p ,r);
			quicksort1(input, p, j-1);
			quicksort1(input, j+1, r);
		}
	}
	
	public static int partition(int[] input, int p, int r) {
		int pivot = input[r];
		while (p < r) {
			while (input[p] < pivot) {
				p++;
			}
			while (input[r] > pivot) {
				r--;
			}
			if (input[p] == input[r]) {
				p++;
			}
			else if (p < r) {
				int tmp = input[p];
				input[p] = input[r];
				input[r] = tmp;
			}
		}
		return r;
	}
}
